### PR TITLE
feat: make logging API more friendly to use

### PR DESCRIPTION
**Pull request by daniel-sanche**
This PR adds a number of small user experience fixes, mostly centered around making the API more forgiving to imperfect inputs

- added support to duplicate certain fields out of logged structs and into the paylolad (eg, severity)
  - https://github.com/googleapis/python-logging/issues/190
- severity values will be auto-converted to upper-case, so they can be parsed properly into enum form
  - This previously caused a crash, when user intention should have been easy to infer
  - https://github.com/googleapis/python-logging/issues/283 
- added tests to ensure pandas dataframes could be logged directly without causing a crash 
  - although directly logging non-string objects is counter to the python logging stdlib API, and is not recommended. Please stringify first instead! 
  - https://github.com/googleapis/python-logging/issues/409
- if users enter a resource in dict format instead of as a full `Resource` object, attempt to parse it into an object instead of failing

### CODE REVIEW

File: google/cloud/logging_v2/logger.py [hunk 1]
```diff
@@ -133,6 +135,20 @@ def _do_log(self, client, _entry_class, payload=None, **kw):
...
+            kw["severity"] = severity.upper()
+
+        if isinstance(kw["resource"], collections.abc.Mapping):
+            # if resource was passed as a dict, attempt to parse it into a
+            # Resource object
+            try:
+                kw["resource"] = Resource(**kw["resource"])
+            except TypeError as e:
+                # dict couldn't be parsed as a Resource
+                raise TypeError(f"invalid resource dict. {e}")
```
Comment (tseaver) [inline google/cloud/logging_v2/logger.py:150]: Rebinding the exception loses information. The default traceback should indicate where the failure occurred, and should be fine for the developer to debug:
```suggestion
            kw["resource"] = Resource(**kw["resource"])
```

Comment (daniel-sanche) [reply google/cloud/logging_v2/logger.py:150]: The problem is the raw error message doesn't make it clear that the resource is the issue:

`TypeError: __new__() missing 2 required positional arguments: 'type' and 'labels'`

That's a good point that we shouldn't lose the context with a fresh exception though. What do you think about [following this pattern](https://stackoverflow.com/a/46091127), to add the extra context?

```
try:
    kw["resource"] = Resource(**kw["resource"])
except TypeError as e:
    raise TypeError("invalid resource dict") from e
```

**Review by losalex — COMMENTED** [review]
LGTM

**Review by minherz — COMMENTED** [review]
nit: if you keep code raising exception if the dictionary is not well formed, it worth adding a test for it as well

**Review by simonz130 — APPROVED** [review]

File: google/cloud/logging_v2/logger.py [hunk 2]
```diff
@@ -133,6 +135,20 @@ def _do_log(self, client, _entry_class, payload=None, **kw):
...
+            kw["severity"] = severity.upper()
+
+        if isinstance(kw["resource"], collections.abc.Mapping):
+            # if resource was passed as a dict, attempt to parse it into a
+            # Resource object
+            try:
+                kw["resource"] = Resource(**kw["resource"])
+            except TypeError as e:
+                # dict couldn't be parsed as a Resource
+                raise TypeError("invalid resource dict") from e
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply google/cloud/logging_v2/logger.py:150]: Before this change, the library would already throw the same exception here for the same inputs. This change attempts to turn certain error-raising inputs into valid inputs. But some inputs will still be invalid

Also, note that [exceptions are used for control flow in Python](https://softwareengineering.stackexchange.com/a/351121), so raising an exception doesn't necessarily mean "crash"

**Comment by daniel-sanche** [comment]
> nit: if you keep code raising exception if the dictionary is not well formed, it worth adding a test for it as well

I did add some system tests around this, but if you think you see any gaps, let me know

File: tests/unit/test_logger.py [hunk 3]
```diff
@@ -379,6 +379,107 @@ def test_log_struct_w_explicit(self):
 
         self.assertEqual(api._write_entries_called_with, (ENTRIES, None, None, None))
 
+    def test_log_struct_inference(self):
+        """
+        if _STRUCT_EXTRACTABLE_FIELDS are present in the struct and not given,
+        they should be used in the LogEntry
```
[the remark this answers is not in the document]
Comment (daniel-sanche) [reply tests/unit/test_logger.py]: I tried to fix it, but it may still be a bit hard to parse. Let me know if you have a suggestion
